index.bs

```diff
@@ -787,14 +718,595 @@ the following list:
+ 1. Set |stream|'s [=WebSocketStream/was ever connected=] to true.
+ 1. Let |extensions| be the [=extensions in use=].
+ 1. Let |protocol| be the [=subprotocol in use=].
+ 1. Let |pullAlgorithm| be an action that [=pulls bytes=] from |stream|.
+ 1. Let |cancelAlgorithm| be an action that [=cancels=] the {{WebSocketStream}} with |stream| and
+    |reason|, given |reason|.
+ 1. Let |readable| be a [=new=] {{ReadableStream}}.
+ 1. [=ReadableStream/Set up=] |readable| with |pullAlgorithm| and |cancelAlgorithm|.
+ 1. Let |writeAlgorithm| be an action that [=writes=] |chunk| to |stream|, given |chunk|.
+ 1. Let |closeAlgorithm| be an action that [=closes=] |stream|.
+ 1. Let |abortAlgorithm| be an action that [=aborts=] |stream| with |reason|, given |reason|.
+ 1. Let |writable| be a [=new=] {{WritableStream}}.
+ 1. [=WritableStream/Set up=] |writable| with |writeAlgorithm|, |closeAlgorithm|, and
+    |abortAlgorithm|.
+ 1. Set |stream|'s [=WebSocketStream/readable stream=] to |readable|.
+ 1. Set |stream|'s [=WebSocketStream/writable stream=] to |writable|.
+ 1. [=Resolve=] |stream|'s [=WebSocketStream/opened promise=] with {{WebSocketOpenInfo}} «[
+    "{{WebSocketOpenInfo/extensions}}" → |extensions|, "{{WebSocketOpenInfo/protocol}}" →
+    |protocol|, "{{WebSocketOpenInfo/readable}}" → |readable|, "{{WebSocketOpenInfo/writable}}" →
+    |writable| ]».
+
+</div>
+
+<hr>
+
+<div algorithm="handle a WebSocketStream message">
+
+When [=a WebSocket message has been received=] for a {{WebSocketStream}} |stream| with type |type|
+and data |data|, the user agent must [=queue a global task=] on the [=WebSocket task source=] given
+|stream|'s [=relevant global object=] to follow these steps: [[!WSP]]
+
+ 1. If [=WebSocket/ready state=] is not {{WebSocket/OPEN}} (1), then return.
+ 1. Let |chunk| be determined by switching on |type|:
+
+  <dl class="switch">
+   : |type| indicates that the data is Text
+   :: a new {{DOMString}} containing |data|
+
+   : |type| indicates that the data is Binary
+   :: a new {{Uint8Array}} object, created in the [=relevant Realm=] of the {{WebSocketStream}}
+      object, whose contents are |data|
+  </dl>
+
+ 1. [=ReadableStream/Enqueue=] |chunk| into |stream|'s [=WebSocketStream/readable stream=].
+ 1. Apply backpressure to the WebSocket.
+
+ <p class=note>Applying backpressure <span class=allow-2119>should</span> result in no new WebSocket
+ messages being handled until the backpressure is released. After an implementation-defined amount
+ of data is buffered, the client will stop reading from the underlying network connection resulting
+ in network-layer backpressure being applied.
+</div>
+
+<hr>
+
+When [=the WebSocket closing handshake is started=] for a {{WebSocketStream}}
+|stream|, the user agent must [=queue a global task=] on the [=WebSocket task source=] given
+|stream|'s [=relevant global object=] to change the |stream|'s [=WebSocketStream/ready state=] to
+{{WebSocket/CLOSING}} if it is not already. [[!WSP]]
```

Comment:
This is a bit ambiguous. Does this mean: queue a task to (if state != CLOSING, set state to CLOSING) or does it mean: if state != CLOSING, queue a task to set state to CLOSING.

These could be observably different due to race conditions.

Possibly it's best to just delete "if it is not already"?